### PR TITLE
hubble-relay: add support for (m)TLS

### PR DESCRIPTION
**Pull request by rolinh**
This PR adds support for TLS and mTLS to Hubble Relay. This allows to configure (m)TLS for the server as well as when establishing connections to peers in the cluster.

Note: This PR depends on #12906.

Ref: #11224

### CODE REVIEW

**Review by kkourt — APPROVED** [review]
LGTM.

File: pkg/hubble/relay/server/option.go [hunk 1]
```diff
@@ -212,3 +212,13 @@ func WithClientMTLSFromCert(ca *x509.CertPool, cert tls.Certificate) Option {
 		MinVersion:   ClientTLSMinVersion,
 	})
 }
+
+// WithInsecureClient disables transport security for connection to Hubble
+// server instances. Transport security is required to WithInsecureClient is
+// set (not recommended).
+func WithInsecureClient() Option {
```
Comment (gandro) [inline pkg/hubble/relay/server/option.go:219]: nit: I find the client/server terminology a bit ambiguous here. Yes, hubble-relay acts as a client, but it also has clients such as UI/CLI that connect to it.

Can we call the client-tls `peer-tls` instead, i.e. name this `WithInsecurePeer` and the options `tls-peer-cert-file` and so forth?

File: hubble-relay/cmd/serve/serve.go [hunk 2]
```diff
@@ -107,11 +111,26 @@ func New(vp *viper.Viper) *cobra.Command {
...
+	)
 	flags.Bool(
 		keyTLSClientDisabled,
 		false,
 		"Disable (m)TLS and allow the connection to Hubble server instances to be over plaintext.",
 	)
+	flags.Bool(
+		keyTLSServerDisabled,
+		false,
+		"Disable (m)TLS for the server and allow clients to connect over plaintext.",
```
Comment (gandro) [inline hubble-relay/cmd/serve/serve.go]:
```suggestion
		"Disable TLS for the server and allow clients to connect over plaintext.",
```

File: pkg/hubble/relay/pool/manager.go [hunk 3]
```diff
@@ -319,8 +319,7 @@ func (m *PeerManager) connect(p *peer, ignoreBackoff bool) {
 	m.opts.log.WithFields(logrus.Fields{
 		"address": p.Address,
 	}).Debugf("Connecting peer %s...", p.Name)
-	//FIXME: provide hostname to ClientConn
-	conn, err := m.opts.clientConnBuilder.ClientConn(p.Address.String(), "")
+	conn, err := m.opts.clientConnBuilder.ClientConn(p.Address.String(), p.TLSServerName)
```
Comment (gandro) [inline pkg/hubble/relay/pool/manager.go:322]: Should we throw a warning if `p.TLSEnabled=true` but hubble-relay was not configured with any certs for TLS for peers?

**Review by gandro — COMMENTED** [review]
Overall looks good to me! Some smaller nits. I can do another pass once the base PR has been merged.

File: pkg/hubble/relay/server/option.go [hunk 4]
```diff
@@ -15,21 +15,40 @@
...
 	"github.com/cilium/cilium/pkg/hubble/relay/defaults"
 	"github.com/cilium/cilium/pkg/hubble/relay/observer"
+	"google.golang.org/grpc/credentials"
+)
+
+const (
+	// ClientTLSMinVersion defines the minimum TLS version that is acceptable
+	// when establishing a connection to a peer. It is not used by
+	// WithClientTLS().
+	ClientTLSMinVersion = tls.VersionTLS13
```
Comment (sayboras) [inline pkg/hubble/relay/server/option.go:33]: do you feel TLS 1.3 is a little bit strict for min version ?

File: pkg/hubble/relay/server/server.go [hunk 5]
```diff
@@ -33,6 +34,16 @@ import (
 	"google.golang.org/grpc/reflection"
 )
 
+var (
+	// ErrNoTransportCredentials is returned when no transport credentials is
+	// set for the server unless WithInsecureServer() is provided.
+	ErrNoTransportCredentials = errors.New("no transport credentials set (use WithInsecureServer() or set credentials)")
+
+	// ErrNoClientTLSConfig is returned when no client TLS config is set
+	// unlesss WithInsecureClient() is provided.
```
Comment (sayboras) [inline pkg/hubble/relay/server/server.go]: nit: unlesss -> unless

File: hubble-relay/cmd/serve/serve.go [hunk 6]
```diff
@@ -117,9 +178,77 @@ func runServe(f flags) error {
...
+	switch {
+	case len(tlsHubbleServerCAPaths) > 0:
+		tlsHubbleServerCAs = x509.NewCertPool()
+		for _, path := range tlsHubbleServerCAPaths {
+			certPEM, err := ioutil.ReadFile(path)
+			if err != nil {
+				return nil, fmt.Errorf("cannot load cert '%s': %s", path, err)
+			}
+			if ok := tlsHubbleServerCAs.AppendCertsFromPEM(certPEM); !ok {
+				return nil, fmt.Errorf("cannot process cert '%s': %s", path, err)
```
Comment (sayboras) [inline hubble-relay/cmd/serve/serve.go]: mostly likely the errr here is nil right ?

File: hubble-relay/cmd/serve/serve.go [hunk 7]
```diff
@@ -117,9 +178,77 @@ func runServe(f flags) error {
...
+
+	var tlsHubbleServerCAs *x509.CertPool
+	tlsHubbleServerCAPaths := vp.GetStringSlice(keyTLSHubbleServerCAFiles)
+	switch {
+	case len(tlsHubbleServerCAPaths) > 0:
+		tlsHubbleServerCAs = x509.NewCertPool()
+		for _, path := range tlsHubbleServerCAPaths {
+			certPEM, err := ioutil.ReadFile(path)
+			if err != nil {
+				return nil, fmt.Errorf("cannot load cert '%s': %s", path, err)
```
Comment (sayboras) [inline hubble-relay/cmd/serve/serve.go:201]: %s -> %w if you want

File: install/kubernetes/cilium/charts/hubble-tls/templates/secret.yaml [hunk 8]
```diff
@@ -45,15 +45,31 @@ data:
 apiVersion: v1
 kind: Secret
 metadata:
-  name: hubble-relay-certs
+  name: hubble-relay-client-certs
```
Comment (sayboras) [inline install/kubernetes/cilium/charts/hubble-tls/templates/secret.yaml:48]: One question for my understanding, seems like any time helm (e.g. upgarde with re-use values) will generate the new values for certs, will hubble-relay get notified and restart ?

File: hubble-relay/cmd/serve/serve.go (hunk 6 above)
Comment (rolinh) [reply hubble-relay/cmd/serve/serve.go]: Indeed, good catch!

File: pkg/hubble/relay/server/option.go (hunk 4 above)
Comment (rolinh) [reply pkg/hubble/relay/server/option.go:33]: I think that for the connection to Hubble server instances, we can be pretty restrictive in what is allowed as we have full control over both servers (Hubble server instances) and clients (Hubble Relay instances). Hubble server instances are not meant to be accessed other than through Hubble Relay.

File: hubble-relay/cmd/serve/serve.go (hunk 7 above)
Comment (rolinh) [reply hubble-relay/cmd/serve/serve.go:201]: I don't think it makes sense to wrap the error here as there is nothing much the caller could do but I can give it more thoughts.

File: install/kubernetes/cilium/charts/hubble-tls/templates/secret.yaml (hunk 8 above)
Comment (rolinh) [reply install/kubernetes/cilium/charts/hubble-tls/templates/secret.yaml:48]: This is a good point. I don't think Hubble Relay will get notified. The best approach to address this is probably to watch for cert files changes (typically via fsnotify) and reload the certs when this happens.
There shouldn't be any reason to restart pods when new certs are generated as there is no point breaking existing connections (it's not a problem if certs are re-generated while connections are active because the TLS certs are only useful for the TLS handshake during which 4 session keys are exchanged to encrypt the communication).
I plan to work on this in a follow-up PR. This one is already pretty big.

File: pkg/hubble/relay/pool/manager.go (hunk 3 above)
Comment (rolinh) [reply pkg/hubble/relay/pool/manager.go:322]: It would be nice to throw a warning but in practice it's a little difficult as the pool manager is not aware whether relay is configured with certs for TLS or not (the `ClientConnBuilder` is set when instantiating a new `Server` and at this point the `Server` is not aware if peers are TLS enabled or not and the pool manager is not aware of how `ClientConn()` establishes the connection (it's an interface)).

TLDR: it would require quite some changes to implement. If you feel it's a must have, I suggest to open an issue and address this later.

**Review by sayboras — APPROVED** [review]
LGTM :100:

File: hubble-relay/cmd/serve/serve.go [hunk 9]
```diff
@@ -117,9 +178,77 @@ func runServe(f flags) error {
...
 }
+
+func buildClientTLSOption(vp *viper.Viper) (server.Option, error) {
+	if vp.GetBool(keyTLSClientDisabled) {
+		return server.WithInsecureClient(), nil
+	}
+
+	var tlsHubbleServerCAs *x509.CertPool
+	tlsHubbleServerCAPaths := vp.GetStringSlice(keyTLSHubbleServerCAFiles)
+	switch {
```
Comment (kaworu) [inline hubble-relay/cmd/serve/serve.go:195]: any reason to use a `switch` for only two cases (i.e. instead of `if`/`else`)?

File: pkg/hubble/peer/types/peer_test.go [hunk 10]
```diff
@@ -149,6 +149,39 @@ func TestFromChangeNotification(t *testing.T) {
...
+				},
+				TLSEnabled: true,
+			},
+		}, {
+			name: "with an IPv4 address and port and TLS enabled with server name",
+			arg: &peerpb.ChangeNotification{
+				Name:    "name",
+				Address: "192.0.2.1:4000",
+				Tls: &peerpb.TLS{
+					ServerName: "*.hubble-grpc.cilium.io",
```
Comment (kaworu) [inline pkg/hubble/peer/types/peer_test.go]: nit: this should rather look like `name.default.hubble-grpc.cilium.io` correct?

File: pkg/hubble/peer/types/peer_test.go [hunk 11]
```diff
@@ -149,6 +149,39 @@ func TestFromChangeNotification(t *testing.T) {
...
+				},
+			},
+			want: &Peer{
+				Name: "name",
+				Address: &net.TCPAddr{
+					IP:   net.ParseIP("192.0.2.1"),
+					Port: 4000,
+				},
+				TLSEnabled:    true,
+				TLSServerName: "*.hubble-grpc.cilium.io",
```
Comment (kaworu) [inline pkg/hubble/peer/types/peer_test.go]: ditto

File: pkg/hubble/relay/pool/client.go [hunk 12]
```diff
@@ -31,11 +33,27 @@ type GRPCClientConnBuilder struct {
...
 // ClientConn implements ClientConnBuilder.ClientConn.
-func (b GRPCClientConnBuilder) ClientConn(target string) (poolTypes.ClientConn, error) {
+func (b GRPCClientConnBuilder) ClientConn(target, hostname string) (poolTypes.ClientConn, error) {
 	ctx, cancel := context.WithTimeout(context.Background(), b.DialTimeout)
 	defer cancel()
-	return grpc.DialContext(ctx, target, b.Options...)
+	opts := make([]grpc.DialOption, len(b.Options))
+	copy(opts, b.Options)
+
+	switch b.TLSConfig {
```
Comment (kaworu) [inline pkg/hubble/relay/pool/client.go:50]: nit: `if`/`else`?

File: pkg/hubble/relay/pool/manager.go (hunk 3 above)
Comment (kaworu) [reply pkg/hubble/relay/pool/manager.go:322]: We can check `m.opts.clientConnBuilder.TLSConfig` right? Alternatively, `ClientConn()` could return an error when it has a TLS config while called with an empty server name.

File: pkg/hubble/relay/pool/manager_test.go [hunk 13]
```diff
@@ -467,6 +467,79 @@ func TestPeerManager(t *testing.T) {
...
+					return &testutils.FakePeerClient{
+						OnNotify: func(_ context.Context, _ *peerpb.NotifyRequest, _ ...grpc.CallOption) (peerpb.Peer_NotifyClient, error) {
+							i := -1
+							cns := []*peerpb.ChangeNotification{
+								{
+									Name:    "one",
+									Address: "192.0.1.1",
+									Type:    peerpb.ChangeNotificationType_PEER_ADDED,
+									Tls: &peerpb.TLS{
+										ServerName: "*.hubble-grpc.cilium.io",
```
Comment (kaworu) [inline pkg/hubble/relay/pool/manager_test.go]: nit:
```suggestion
										ServerName: "one.default.hubble-grpc.cilium.io",
```

File: pkg/hubble/relay/pool/manager_test.go [hunk 14]
```diff
@@ -467,6 +467,79 @@ func TestPeerManager(t *testing.T) {
...
+									},
+								}, {
+									Name:    "two",
+									Address: "192.0.1.2",
+									Type:    peerpb.ChangeNotificationType_PEER_ADDED,
+									Tls: &peerpb.TLS{
+										ServerName: "*.hubble-grpc.cilium.io",
```
Comment (kaworu) [inline pkg/hubble/relay/pool/manager_test.go]: nit:
```suggestion
										ServerName: "two.default.hubble-grpc.cilium.io",
```

File: pkg/hubble/relay/pool/manager_test.go [hunk 15]
```diff
@@ -467,6 +467,79 @@ func TestPeerManager(t *testing.T) {
...
+									},
+								}, {
+									Name:    "one",
+									Address: "192.0.1.1",
+									Type:    peerpb.ChangeNotificationType_PEER_DELETED,
+									Tls: &peerpb.TLS{
+										ServerName: "*.hubble-grpc.cilium.io",
```
Comment (kaworu) [inline pkg/hubble/relay/pool/manager_test.go]: nit:
```suggestion
										ServerName: "one.default.hubble-grpc.cilium.io",
```

File: pkg/hubble/relay/pool/manager_test.go [hunk 16]
```diff
@@ -467,6 +467,79 @@ func TestPeerManager(t *testing.T) {
...
+				peers: []poolTypes.Peer{
+					{
+						Peer: peerTypes.Peer{
+							Name: "two",
+							Address: &net.TCPAddr{
+								IP:   net.ParseIP("192.0.1.2"),
+								Port: defaults.ServerPort,
+							},
+							TLSEnabled:    true,
+							TLSServerName: "*.hubble-grpc.cilium.io",
```
Comment (kaworu) [inline pkg/hubble/relay/pool/manager_test.go]:
```suggestion
							TLSServerName: "two.default.hubble-grpc.cilium.io",
```

File: pkg/hubble/relay/server/option.go [hunk 17]
```diff
@@ -125,3 +144,81 @@ func WithErrorAggregationWindow(d time.Duration) Option {
...
+	return WithTLS(&tls.Config{
+		Certificates: []tls.Certificate{cert},
+		ClientCAs:    clientCAs,
+		ClientAuth:   tls.RequireAndVerifyClientCert,
+		MinVersion:   ServerTLSMinVersion,
+	})
+}
+
+// WithClientTLS sets the transport credentials for connecting to peers based
+// on the provided TLS configuration>
```
Comment (kaworu) [inline pkg/hubble/relay/server/option.go]:
```suggestion
// on the provided TLS configuration.
```

File: pkg/hubble/relay/server/option.go [hunk 18]
```diff
@@ -125,3 +144,81 @@ func WithErrorAggregationWindow(d time.Duration) Option {
...
+func WithClientTLSFromCert(ca *x509.CertPool) Option {
+	return WithClientTLS(&tls.Config{
+		RootCAs:    ca,
+		MinVersion: ClientTLSMinVersion,
+	})
+}
+
+// WithClientMTLSFromCert constructs mutual TLS (mTLS) credentials from ca and
+// cert. The certificate is presented to the peer and the provided root
+// certificate authority is used to validate connections to peers.
```
Comment (kaworu) [inline pkg/hubble/relay/server/option.go:207]:
```suggestion
// cert. The certificate is presented to the peers, and the provided root
// certificate authority is used to validate certificates presented by the peers.
```

File: pkg/hubble/relay/server/option.go [hunk 19]
```diff
@@ -125,3 +144,81 @@ func WithErrorAggregationWindow(d time.Duration) Option {
...
+func WithClientMTLSFromCert(ca *x509.CertPool, cert tls.Certificate) Option {
+	return WithClientTLS(&tls.Config{
+		Certificates: []tls.Certificate{cert},
+		RootCAs:      ca,
+		MinVersion:   ClientTLSMinVersion,
+	})
+}
+
+// WithInsecureClient disables transport security for connection to Hubble
+// server instances. Transport security is required to WithInsecureClient is
```
Comment (kaworu) [inline pkg/hubble/relay/server/option.go:217]:
```suggestion
// server instances. Transport security is required unless WithInsecureClient is
```

**Review by nebril — APPROVED** [review]

**Review by kaworu — APPROVED** [review]
Awesome PR! Only have some nits.

File: pkg/hubble/peer/types/peer_test.go (hunk 10 above)
Comment (rolinh) [reply pkg/hubble/peer/types/peer_test.go]: Indeed. Will fix.

**Review by gandro — APPROVED** [review]

File: pkg/hubble/relay/server/option.go (hunk 1 above)
Comment (rolinh) [reply pkg/hubble/relay/server/option.go:219]: That's a valid point. I'll try to come up with a terminology that is consistent and not confusing and will amend the PR.

File: pkg/hubble/relay/pool/manager.go (hunk 3 above)
Comment (rolinh) [reply pkg/hubble/relay/pool/manager.go:322]: > We can check m.opts.clientConnBuilder.TLSConfig right?

I'm not really sure what you are trying to imply here.  The current code uses the `GRPCClientConnBuilder` as the implementation of the `ClientConnBuilder` interface. This implementation opts to connect using TLS if a TLS config is provided and uses plaintext otherwise.

I think that what @gandro is pointing out is that the peer service already reports whether a peer is configured with TLS or not (see `Peer.TLSEnabled`) but this information is discarded by the pool manager which, by abstraction, knows nothing about how the connection to peers is established. In order to change this, the `ClientConnBuilder` interface should be modified such that the caller defines whether TLS is required or not.

> Alternatively, ClientConn() could return an error when it has a TLS config while called with an empty server name.

I think that this is what happens in practice currently as the TLS handshake will fail as the Server Name will not match the certificate's.

Comment (kaworu) [reply pkg/hubble/relay/pool/manager.go:322]: Let's follow up in #13087

**Comment by Rolinh** [comment]
test-me-please

File: pkg/hubble/relay/server/option.go (hunk 1 above)
Comment (rolinh) [reply pkg/hubble/relay/server/option.go:219]: @kAworu opened #13094

**Review by nathanjsweet — APPROVED** [review]

File: install/kubernetes/cilium/charts/hubble-relay/templates/configmap.yaml [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+{{- end }}
+{{- if .Values.retryTimeout }}
+    retry-timeout: {{ .Values.retryTimeout }}
+{{- end }}
+{{- if .Values.sortBufferLenMax }}
+    sort-buffer-len-max: {{ .Values.sortBufferLenMax }}
+{{- end }}
+{{- if .Values.sortBufferDrainTimeout }}
+    sort-buffer-drain-timeout: {{ .Values.sortBufferDrainTimeout }}
+{{- end }}
```
Comment (aanm) [inline install/kubernetes/cilium/charts/hubble-relay/templates/configmap.yaml:25]: We should have the same implementation when setting up values in the config map as done in Cilium for consistency.
